File: BTFViewer/web/src/utils/evidenceNav.js
/**
 * Universal Evidence Navigation (Step 2).
 * Lockstep with btf_viewer_pkg/evidence_nav.py.
 *
 * Show Evidence centers the Timeline, places/reuses one Evidence cursor, and
 * highlights the related Task — without silently changing Scope or Filters.
 */
import { bestFindingScope } from './uxExplore.js'

export const EVIDENCE_GLYPH = '\u2197' // ↗

/** Tooltip for Evidence-affordance cells (Statistics / Findings / Compare). */
export const EVIDENCE_TOOLTIP = 'Jump to Evidence (does not change Scope)'

const TIME_TOKEN_RE = /(?:jump:)?(\d+(?:\.\d+)?)\s*(ns|us|µs|μs|ms|s)?/gi

const UNIT_NS = {
  '': 1,
  ns: 1,
  us: 1e3,
  'µs': 1e3,
  'μs': 1e3,
  ms: 1e6,
  s: 1e9,
}

export function parseEvidenceTimestamps(...texts) {
  const out = []
  const seen = new Set()
  for (const text of texts) {
    const blob = String(text || '')
    TIME_TOKEN_RE.lastIndex = 0
    let m
    while ((m = TIME_TOKEN_RE.exec(blob)) !== null) {
      const value = Number(m[1])
      if (!Number.isFinite(value)) continue
      const unit = String(m[2] || '').toLowerCase()
      const raw = m[0]
      let scale = UNIT_NS[unit] ?? 1
      if (!unit && !/jump:/i.test(raw)) {
        if (value < 1e6) continue
        scale = 1
      }
      const ns = Math.trunc(value * scale)
      if (ns < 0 || seen.has(ns)) continue
      seen.add(ns)
      out.push(ns)
    }
  }
  return out
}

export function resolveFindingEvidence(finding, events, timeMin, timeMax) {
  if (!finding || typeof finding !== 'object') {
    return {
      ok: false,
      ns: null,
      task: '',
      mk: '',
      note: '',
      multi: false,
      reason: 'No finding selected',
    }
  }

  const times = []
  for (const ev of finding.evidence || []) {
    if (!ev || typeof ev !== 'object') continue
    for (const key of ['time', 'start', 'stop', 'ns']) {
      if (ev[key] == null) continue
      const v = Number(ev[key])
      if (Number.isFinite(v)) times.push(v)
    }
  }
  times.push(
    ...parseEvidenceTimestamps(
      finding.evidence_text || '',
      finding.title || '',
      finding.text || '',
    ),
  )

  const scope = bestFindingScope(finding, events, timeMin, timeMax)
  let task = String(scope?.task || finding.task || '').trim()
  let mk = String(scope?.mk || '').trim()
  const multi = new Set(times.map(t => Math.trunc(t))).size > 1

  let ns = null
  let note = ''
  if (times.length) {
    ns = Math.trunc(Math.max(...times))
    note = multi
      ? 'Representative (latest) Evidence among multiple samples'
      : 'Evidence timestamp from finding'
  } else if (scope) {
    const lo = Math.trunc(scope.lo || 0)
    const hi = Math.trunc(scope.hi || lo)
    ns = Math.trunc((lo + hi) / 2)
    note = scope.reason || 'Representative moment in finding episode'
    task = task || String(scope.task || '')
    mk = mk || String(scope.mk || '')
  }

  if (ns == null) {
    return {
      ok: false,
      ns: null,
      task,
      mk,
      note: '',
      multi: false,
      reason: 'No locatable Timeline Evidence for this finding',
    }
  }

  const tmin = Math.trunc(timeMin || 0)
  const tmax = Math.trunc(timeMax || 0)
  if (tmax > tmin) ns = Math.max(tmin, Math.min(tmax, ns))

  return {
    ok: true,
    ns,
    task,
    mk,
    note,
    multi,
    reason: '',
  }
}

export function resolveTimestampEvidence(
  ns,
  { task = '', mk = '', note = '', timeMin = 0, timeMax = 0 } = {},
) {
  if (ns == null || ns === '') {
    return {
      ok: false,
      ns: null,
      task: task || '',
      mk: mk || '',
      note: '',
      multi: false,
      reason: 'Evidence timestamp is missing or invalid',
    }
  }
  const value = Number(ns)
  if (!Number.isFinite(value)) {
    return {
      ok: false,
      ns: null,
      task: task || '',
      mk: mk || '',
      note: '',
      multi: false,
      reason: 'Evidence timestamp is missing or invalid',
    }
  }
  let out = Math.trunc(value)
  const tmin = Math.trunc(timeMin || 0)
  const tmax = Math.trunc(timeMax || 0)
  if (tmax > tmin) out = Math.max(tmin, Math.min(tmax, out))
  return {
    ok: true,
    ns: out,
    task: String(task || ''),
    mk: String(mk || ''),
    note: note || 'Evidence timestamp',
    multi: false,
    reason: '',
  }
}

File: BTFViewer/btf_viewer_pkg/evidence_nav.py
"""Universal Evidence Navigation (Step 2).

One shared contract for jumping from analytical results (Findings, Statistics,
AI, Migration, Compare) to concrete Timeline Evidence.

Rules (lockstep with ``web/src/utils/evidenceNav.js``):

* Center the Timeline on the Evidence timestamp.
* Place or reuse a single Evidence cursor at that time (does **not** redefine
  Scope / C1–Cn analysis window).
* Highlight the related Task when known.
* Never silently change Scope or Filters.
* When several representative events exist, prefer the longest / worst in the
  finding's evidence set (same ranking as ``best_finding_scope``).
* When no exact Evidence can be located, return ``ok=False`` with a reason —
  callers show a status message instead of inventing a jump.
"""
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Sequence

from .ux_explore import best_finding_scope  # noqa: F401 — re-exported for callers

# Visible affordance glyph used next to actionable Evidence timestamps.
EVIDENCE_GLYPH = "\u2197"  # ↗

# Tooltip for Evidence-affordance cells (Statistics / Findings / Compare).
EVIDENCE_TOOLTIP = "Jump to Evidence (does not change Scope or Filters)"

_TIME_TOKEN_RE = re.compile(
    r"(?:jump:)?(\d+(?:\.\d+)?)\s*(ns|us|µs|μs|ms|s)?",
    re.IGNORECASE,
)

_UNIT_NS = {
    "": 1.0,
    "ns": 1.0,
    "us": 1_000.0,
    "µs": 1_000.0,
    "μs": 1_000.0,
    "ms": 1_000_000.0,
    "s": 1_000_000_000.0,
}


def parse_evidence_timestamps(*texts: str) -> List[int]:
    """Extract candidate timestamps (ns) from free-form Evidence strings."""
    out: List[int] = []
    seen = set()
    for text in texts:
        blob = str(text or "")
        for m in _TIME_TOKEN_RE.finditer(blob):
            try:
                value = float(m.group(1))
            except (TypeError, ValueError):
                continue
            unit = (m.group(2) or "").lower()
            # Bare integers in evidence_text are often counts, not times —
            # only accept unitless values when tagged with jump: or clearly large.
            raw = m.group(0)
            if not unit and "jump:" not in raw.lower():
                if value < 1_000_000:  # smaller than 1 ms as bare ns is ambiguous
                    continue
                scale = 1.0
            else:
                scale = _UNIT_NS.get(unit, 1.0)
            ns = int(value * scale)
            if ns in seen or ns < 0:
                continue
            seen.add(ns)
            out.append(ns)
    return out


def resolve_finding_evidence(
    finding: Optional[dict],
    events: Sequence[dict],
    time_min: float,
    time_max: float,
) -> Dict[str, Any]:
    """Resolve Timeline Evidence for an Analysis Finding without changing Scope.

    Returns a dict:
      ok, ns, task, mk, note, multi, reason
    """
    if not isinstance(finding, dict):
        return {
            "ok": False,
            "ns": None,
            "task": "",
            "mk": "",
            "note": "",
            "multi": False,
            "reason": "No finding selected",
        }

    times: List[float] = []
    for ev in finding.get("evidence") or []:
        if not isinstance(ev, dict):
            continue
        for key in ("time", "start", "stop", "ns"):
            try:
                if ev.get(key) is not None:
                    times.append(float(ev[key]))
            except (TypeError, ValueError):
                continue
    times.extend(
        float(t)
        for t in parse_evidence_timestamps(
            str(finding.get("evidence_text") or ""),
            str(finding.get("title") or ""),
            str(finding.get("text") or ""),
        )
    )

    scope = best_finding_scope(finding, events, time_min, time_max)
    task = str((scope or {}).get("task") or finding.get("task") or "").strip()
    mk = str((scope or {}).get("mk") or "").strip()
    multi = len(set(int(t) for t in times)) > 1

    ns: Optional[int] = None
    note = ""
    if times:
        # Prefer the latest evidence sample when several exist (tail / worst).
        ns = int(max(times))
        note = "Evidence timestamp from finding"
        if multi:
            note = "Representative (latest) Evidence among multiple samples"
    elif scope is not None:
        # Fall back to mid-point of recommended episode — still no Scope change.
        lo = int(scope.get("lo") or 0)
        hi = int(scope.get("hi") or lo)
        ns = (lo + hi) // 2
        note = str(scope.get("reason") or "Representative moment in finding episode")
        task = task or str(scope.get("task") or "")
        mk = mk or str(scope.get("mk") or "")

    if ns is None:
        return {
            "ok": False,
            "ns": None,
            "task": task,
            "mk": mk,
            "note": "",
            "multi": False,
            "reason": "No locatable Timeline Evidence for this finding",
        }

    tmin = int(time_min)
    tmax = int(time_max)
    if tmax > tmin:
        ns = max(tmin, min(tmax, ns))

    return {
        "ok": True,
        "ns": int(ns),
        "task": task,
        "mk": mk,
        "note": note,
        "multi": multi,
        "reason": "",
    }


def resolve_timestamp_evidence(
    ns: Any,
    *,
    task: str = "",
    mk: str = "",
    note: str = "",
    time_min: float = 0,
    time_max: float = 0,
) -> Dict[str, Any]:
    """Wrap a concrete timestamp (Statistics cell / AI jump / Compare delta)."""
    try:
        value = int(ns)
    except (TypeError, ValueError):
        return {
            "ok": False,
            "ns": None,
            "task": task or "",
            "mk": mk or "",
            "note": "",
            "multi": False,
            "reason": "Evidence timestamp is missing or invalid",
        }
    tmin = int(time_min or 0)
    tmax = int(time_max or 0)
    if tmax > tmin:
        value = max(tmin, min(tmax, value))
    return {
        "ok": True,
        "ns": value,
        "task": str(task or ""),
        "mk": str(mk or ""),
        "note": note or "Evidence timestamp",
        "multi": False,
        "reason": "",
    }
